Add SNMP trap configuration parameters (#449)
infrawatch/documentation: doc-Service-Telemetry-Framework/modules/proc_configuring-snmp-traps.adoc

[id="configuring-snmp-traps_{context}"]
= Configuring SNMP traps

[role="_abstract"]

> You can integrate {Project} ({ProjectShort}) with an existing infrastructure monitoring platform that receives notifications through SNMP traps. To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters.

Reviewer: You can't have a module with no short introduction (for seo purposes)

> You can integrate {Project} ({ProjectShort}) with an existing infrastructure monitoring platform that receives notifications through SNMP traps. To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters.

Reviewer: ```suggestion
To enable SNMP traps, modify the `ServiceTelemetry` object and configure the `snmpTraps` parameters. 
```
You need some kind of intro to the module, can't be blank.

Reviewer: @mickogeary These suggested edits are here for you rather than for Leif, commit or ignore, whatever you see fit.

Reviewer 2: It's telling me "Applying suggestions on deleted lines is currently not supported. "

Author: Same here. I will look and see if this line still exists and move the suggested change there.

Author: Moved to other section. Applied change there.

> * Know the IP address or hostname of the SNMP trap receiver where you want to send the alerts

Reviewer 2: ```suggestion
* Ensure that you know the IP address or hostname of the SNMP trap receiver where you want to send the alerts to.
```

----

. Ensure that you set the value of `target` to the IP address or hostname of the SNMP trap receiver.

.Additional Information

For more information about available parameters for `snmpTraps`, see xref:configuration-parameters-for-snmptraps_assembly-advanced-features[].

[id="creating-alerts-for-snmp-traps_{context}"]
= Creating alerts for SNMP traps

> You can create alerts that are configured for delivery via SNMP traps by adding labels that are parsed by the prometheus-webhook-snmp middleware to define the trap information and delivered object identifiers (OID). Defining alerts in this way is only required if you need to override the default OID SNMP traps are delivered to or you need to override the trap prefix OID.

Author: I want to rework this a bit. The final sentence isn't entirely correct, as overriding the severity could be warranted. An example of why you might want to do this could be to have an alarm of a component use an OID to simulate an existing MIB database. I could see during creation of some alarms there may be some other more well known MIBs to alarm from. As it stands today though, the implementation will be just sending general alarms.

.Procedure

. Log in to {OpenShift}.

. Change to the `service-telemetry` namespace:
+
[source,bash]
----
$ oc project service-telemetry
----

. Create a `PrometheusRule` object that contains the alert rule and an `oid` label that contains the SNMP trap OID override value:
+
[source,bash]
----
$ oc apply -f - <<EOF
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  creationTimestamp: null
  labels:
    prometheus: default
    role: alert-rules
  name: prometheus-alarm-rules-snmp
  namespace: service-telemetry
spec:
  groups:
    - name: ./openstack.rules
      rules:
        - alert: Collectd metrics receive rate is zero
          expr: rate(sg_total_collectd_msg_received_count[1m]) == 0
          labels:

>             oid: 1.3.6.1.4.1.50495.15.1.2.1

Reviewer 3: How does this affect the "sub objects" you show above? If you change this, I assume it changes the top-line OID that the trap is identified as, but for example, will it still include `<trapOidPrefix>.1.1.2` for status in the body?

Author: That is my understanding as well. The ability to change the OID in the alarm definition might seem useful, but it might not be practical because of this. I haven't found an ability to manipulate the entire trap so you'd still have to manipulate your MIB database to support the "prometheus" object within the custom transport OID.

----

. Ensure that you set the value of `target` to the IP address or hostname of the SNMP trap receiver.

.Additional Information

For more information about available parameters for `snmpTraps`, see xref:configuration-parameters-for-snmptraps_assembly-advanced-features[].

[id="creating-alerts-for-snmp-traps_{context}"]
= Creating alerts for SNMP traps

> You can create alerts that are configured for delivery via SNMP traps by adding labels that are parsed by the prometheus-webhook-snmp middleware to define the trap information and delivered object identifiers (OID). Adding the `oid` or `severity` labels is only required if you need to change the default values for a particular alert definition.

Reviewer 2: IBM style guide recommends against using "via", I guess because it is Latin. Can just say "by".

Reviewer 2: ```suggestion
You can create alerts that are configured for delivery by SNMP traps by adding labels that are parsed by the prometheus-webhook-snmp middleware to define the trap information and delivered object identifiers (OID). Adding the `oid` or `severity` labels is only required if you need to change the default values for a particular alert definition.
```

> NOTE:: When the `oid` label is set, the top-level SNMP trap OID is changed, but the sub-OIDs remain defined by the global `trapOidPrefix` value plus the child OID values `.1.1.1` through `.1.1.9`. For more information about the MIB definition, see xref:overview-of-the-mib-definition_{context}[].

Reviewer 2: Maybe: "When you set the `oid` label, the top-level SNMP trap OID changes,"

Reviewer 2: ```suggestion
NOTE:: When you set the oid label, the top-level SNMP trap OID changes, but the sub-OIDs remain defined by the global `trapOidPrefix` value plus the child OID values `.1.1.1` through `.1.1.9`. For more information about the MIB definition, see xref:overview-of-the-mib-definition_{context}[].
```
